File: test/commands/search.test.js
import {
  runClangdQuery,
  assert,
  assertContains,
  assertMatches,
  countOccurrences,
  waitForDaemonReady,
} from '../test-helpers.js';

console.log('Testing search command...\n');

// Ensure daemon is ready before running tests
console.log('Waiting for daemon to be ready...');
await waitForDaemonReady();

// Test 1: Search for GameObject class
console.log('Test 1: Search for GameObject class');
const result1 = await runClangdQuery(['search', 'GameObject']);
assert(result1.exitCode === 0, 'Command should succeed');
assertContains(result1.stdout, 'class game_engine::GameObject');
assertContains(result1.stdout, 'game_engine::GameObject::GameObject');
assertContains(result1.stdout, 'game_engine::Engine::CreateGameObject');
assertContains(result1.stdout, 'game_engine::Engine::DestroyGameObject');
assertContains(result1.stdout, 'game_engine::Engine::GetGameObject');
assertContains(result1.stdout, 'game_engine::GameObject::~GameObject');

console.log('✓ Test 1 passed\n');

// Test 2: Search for Character with inheritance
console.log('Test 2: Search for Character class and related symbols');
const result2 = await runClangdQuery(['search', 'Character']);
assert(result2.exitCode === 0, 'Command should succeed');
assertContains(result2.stdout, 'class game_engine::Character');
// Should find Character class and its constructors/destructors
assertMatches(result2.stdout, /Character.*character\.h/);
console.log('✓ Test 2 passed\n');

// Test 3: Search with limit flag
console.log('Test 3: Search with --limit flag');
const result3 = await runClangdQuery(['search', 'Update', '--limit', '3']);
assert(result3.exitCode === 0, 'Command should succeed');
// Count the number of result lines (excluding header)
const lines3 = result3.stdout.split('\n').filter(line => line.includes(' at '));
assert(lines3.length <= 3, `Should return at most 3 results, got ${lines3.length}`);
console.log('✓ Test 3 passed\n');

// Test 4: Search for non-existent symbol
console.log('Test 4: Search for non-existent symbol');
const result4 = await runClangdQuery(['search', 'NonExistentSymbol']);
assert(result4.exitCode === 0, 'Command should succeed even with no results');
assertContains(result4.stdout, 'No symbols found matching "NonExistentSymbol"');
console.log('✓ Test 4 passed\n');

// Test 5: Search for factory methods
console.log('Test 5: Search for factory methods');
const result5 = await runClangdQuery(['search', 'Create']);
assert(result5.exitCode === 0, 'Command should succeed');
// Should find factory Create methods
assertContains(result5.stdout, 'game_engine::Factory::Create');
assertContains(result5.stdout, 'game_engine::EnemyFactory::CreateEnemy');
assertContains(result5.stdout, 'game_engine::Engine::CreateGameObject');
console.log('✓ Test 5 passed\n');

// Test 6: Case-insensitive fuzzy search
console.log('Test 6: Fuzzy search test');
const result6 = await runClangdQuery(['search', 'updatable']);
assert(result6.exitCode === 0, 'Command should succeed');
// Should find Updatable interface even with lowercase search
assertContains(result6.stdout, 'Updatable');
console.log('✓ Test 6 passed\n');

console.log('All search tests passed! ✓');

File: go/test/search_test.go
package test

import (
	"strings"
	"testing"
)

func TestSearchCommand(t *testing.T) {
	tc := GetTestContext(t)

	t.Run("Search for GameObject class", func(t *testing.T) {
		result := tc.RunCommand("search", "GameObject")
		tc.AssertExitCode(result, 0)

		tc.AssertContains(result.Stdout,
			// Note this string starts with "Found X symbols", but the tests are flaky,
			// sometimes matching 7 symbols, sometimes 8. So we just omit this part
			// of the string. The important thing is that the GameObject results
			// are first.
			" matching \"GameObject\":\n"+
				"\n"+
				"- `class game_engine::GameObject` at include/core/game_object.h:22:7\n"+
				"- `game_engine::GameObject::GameObject` at src/core/game_object.cpp:12:13\n"+
				"- `game_engine::Engine::game_objects_` at include/core/engine.h:120:44\n"+
				"- `game_engine::Engine::CreateGameObject` at src/core/engine.cpp:136:37\n"+
				"- `game_engine::Engine::DestroyGameObject` at src/core/engine.cpp:142:14\n"+
				"- `game_engine::Engine::GetGameObjects` at include/core/engine.h:94:51\n")
		tc.AssertContains(result.Stdout, "game_engine::GameObject::~GameObject")
	})

	t.Run("Search for Character class and related symbols", func(t *testing.T) {
		result := tc.RunCommand("search", "Character")
		tc.AssertExitCode(result, 0)
		tc.AssertContains(result.Stdout, "class game_engine::Character")
		// Should find Character class in character.h
		if !strings.Contains(result.Stdout, "Character") || !strings.Contains(result.Stdout, "character.h") {
			t.Errorf("Expected to find Character in character.h")
		}
	})

	t.Run("Search with limit flag", func(t *testing.T) {
		result := tc.RunCommand("search", "Update", "--limit", "3")
		tc.AssertExitCode(result, 0)
		// Count the number of result lines (those containing ' at ')
		lines := strings.Split(result.Stdout, "\n")
		resultCount := 0
		for _, line := range lines {
			if strings.Contains(line, " at ") {
				resultCount++
			}
		}
		if resultCount > 3 {
			t.Errorf("Expected at most 3 results with --limit 3, got %d", resultCount)
		}
	})

	t.Run("Search for non-existent symbol", func(t *testing.T) {
		result := tc.RunCommand("search", "NonExistentSymbol")
		tc.AssertExitCode(result, 0)
		tc.AssertContains(result.Stdout, "No symbols found matching \"NonExistentSymbol\"")
	})

	t.Run("Search for factory methods", func(t *testing.T) {
		result := tc.RunCommand("search", "Create")
		tc.AssertExitCode(result, 0)
		tc.AssertContains(result.Stdout, "game_engine::Factory::Create")
		tc.AssertContains(result.Stdout, "game_engine::EnemyFactory::CreateEnemy")
		tc.AssertContains(result.Stdout, "game_engine::Engine::CreateGameObject")
	})

	t.Run("Fuzzy search test", func(t *testing.T) {
		result := tc.RunCommand("search", "updatable")
		tc.AssertExitCode(result, 0)
		// Should find Updatable interface even with lowercase search
		tc.AssertContains(result.Stdout, "Updatable")
	})
}
